front50-web/src/main/java/com/netflix/spinnaker/front50/controllers/v2/ProjectsController.java

```diff
@@ -0,0 +1,244 @@
+      projectDAO.findByName(project.getName());
+      throw new InvalidRequestException(
+          format("A Project named '%s' already exists", project.getName()));
+    } catch (NotFoundException ignored) {
+    }
+
+    return projectDAO.create(project.getId(), project);
+  }
+
+  private Set<Project> filter(Collection<Project> projects, Map<String, String> attributes) {
+    Map<String, String> normalizedAttributes = new HashMap<>();
+    for (Map.Entry<String, String> attr : attributes.entrySet()) {
+      if (!Strings.isNullOrEmpty(attr.getValue())) {
+        normalizedAttributes.put(attr.getKey().toLowerCase(), attr.getValue());
+      }
+    }
+
+    Set<Project> items = new HashSet<>();
+
+    if (normalizedAttributes.containsKey("applications")) {
+      List<String> applications =
+          COMMA_SPLITTER.splitToList(normalizedAttributes.get("applications"));
+      // TODO(rz): LOL converting Groovy with elvis operators everywhere. :grimacing:
+      //  Do you know what's going on here? I certainly don't. Good thing there's tests?
+      items =
+          projects.stream()
+              .filter(
+                  project ->
+                      Optional.ofNullable(project.getConfig().getApplications())
+                              .map(
+                                  projectApps ->
+                                      projectApps.stream()
+                                          .anyMatch(
+                                              app -> projectApplicationMatches(app, applications)))
+                              .orElse(false)
+                          || Optional.ofNullable(project.getConfig().getClusters())
+                              .map(
+                                  projectClusters ->
+                                      projectClusters.stream()
+                                          .anyMatch(
+                                              cluster ->
+                                                  clusterHasMatchingApplication(
+                                                      cluster, applications)))
+                              .orElse(false))
+              .collect(Collectors.toSet());
+      normalizedAttributes.remove("applications");
+    }
+
+    items.addAll(
+        projects.stream()
+            .filter(
+                project ->
+                    normalizedAttributes.entrySet().stream()
+                        .anyMatch(
+                            it ->
+                                FUZZY_SEARCH_PREDICATE.test(
+                                    new FuzzySearch(project, it.getKey(), it.getValue()))))
+            .collect(Collectors.toSet()));
+
+    return new TreeSet<>(items)
+        .stream()
+            .sorted(
+                (a, b) ->
+                    SearchUtils.score(b, normalizedAttributes)
+                        - SearchUtils.score(a, normalizedAttributes))
+            .collect(Collectors.toCollection(LinkedHashSet::new));
```

Comment:
I broke this monstrosity up into a few blocks and methods. I'm happy to break it up further.

Reply:
yikes... but yeah it looks like you captured the original essencé d'groovy - nice job